File: cipher-codes/PolybiusCipher.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class PolybiusCipher {

    public static Map<Character, String> createPolybiusGrid(String key) {
        Map<Character, String> grid = new HashMap<>();
        int index = 0;
        for (int row = 1; row <= 5; row++) {
            for (int col = 1; col <= 5; col++) {
                if (index < key.length()) {
                    grid.put(key.charAt(index), String.valueOf(row) + col);
                    index++;
                }
            }
        }
        return grid;
    }

    public static Map<String, Character> reversePolybiusGrid(Map<Character, String> grid) {
        Map<String, Character> reverseGrid = new HashMap<>();
        for (Map.Entry<Character, String> entry : grid.entrySet()) {
            reverseGrid.put(entry.getValue(), entry.getKey());
        }
        return reverseGrid;
    }

    public static String encryptPolybius(String plaintext, Map<Character, String> grid) {
        StringBuilder ciphertext = new StringBuilder();
        for (char c : plaintext.toUpperCase().toCharArray()) {
            if (grid.containsKey(c)) {
                ciphertext.append(grid.get(c));
            } else {
                ciphertext.append(c);
            }
        }
        return ciphertext.toString();
    }

    public static String decryptPolybius(String ciphertext, Map<String, Character> reverseGrid) {
        StringBuilder plaintext = new StringBuilder();
        int i = 0;
        while (i < ciphertext.length()) {
            if (i + 1 < ciphertext.length() && Character.isDigit(ciphertext.charAt(i)) && Character.isDigit(ciphertext.charAt(i + 1))) {
                String key = "" + ciphertext.charAt(i) + ciphertext.charAt(i + 1);
                plaintext.append(reverseGrid.getOrDefault(key, ' '));
                i += 2;
            } else {
                plaintext.append(ciphertext.charAt(i));
                i++;
            }
        }
        return plaintext.toString();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the characters of the Polybius grid (25 unique characters): ");
        String key = scanner.nextLine().toUpperCase();

        if (key.length() != 25 || key.chars().distinct().count() != 25) {
            System.out.println("Error: The key must contain 25 unique characters.");
            scanner.close();
            return;
        }

        Map<Character, String> polybiusGrid = createPolybiusGrid(key);
        Map<String, Character> reverseGrid = reversePolybiusGrid(polybiusGrid);

        System.out.print("Do you want to encrypt or decrypt? (e/d): ");
        String mode = scanner.nextLine().trim().toLowerCase();

        if (mode.equals("e")) {
            System.out.print("Enter the plaintext message: ");
            String plaintext = scanner.nextLine();
            String encryptedMessage = encryptPolybius(plaintext, polybiusGrid);
            System.out.println("Encrypted message: " + encryptedMessage);
        } else if (mode.equals("d")) {
            System.out.print("Enter the ciphertext message: ");
            String ciphertext = scanner.nextLine();
            String decryptedMessage = decryptPolybius(ciphertext, reverseGrid);
            System.out.println("Decrypted message: " + decryptedMessage);
        } else {
            System.out.println("Invalid option. Please choose 'e' to encrypt or 'd' to decrypt.");
        }

        scanner.close();
    }
}

File: cipher-codes/PolybiusCipher.py
def create_polybius_grid(key):
    grid = {}
    index = 0
    for row in range(5):
        for col in range(5):
            if index < len(key):
                grid[key[index]] = (row + 1, col + 1)
                index += 1
    return grid

def reverse_polybius_grid(grid):
    reverse_grid = {}
    for letter, (row, col) in grid.items():
        reverse_grid[(row, col)] = letter
    return reverse_grid

def encrypt_polybius(plaintext, grid):
    ciphertext = ""
    for char in plaintext.upper():
        if char in grid:
            row, col = grid[char]
            ciphertext += str(row) + str(col)
        else:
            ciphertext += char
    return ciphertext

def decrypt_polybius(ciphertext, reverse_grid):
    plaintext = ""
    i = 0
    while i < len(ciphertext):
        if ciphertext[i].isdigit() and (i + 1) < len(ciphertext) and ciphertext[i + 1].isdigit():
            row = int(ciphertext[i])
            col = int(ciphertext[i + 1])
            plaintext += reverse_grid.get((row, col), '')
            i += 2
        else:
            plaintext += ciphertext[i]
            i += 1
    return plaintext

def main():
    key = input("Enter the characters of the Polybius grid (25 unique characters): ").upper()
    
    if len(set(key)) != 25:
        print("Error: The key must contain 25 unique characters.")
        return

    polybius_grid = create_polybius_grid(key)
    reverse_grid = reverse_polybius_grid(polybius_grid)
    
    mode = input("Do you want to encrypt or decrypt? (e/d): ").strip().lower()
    
    if mode == 'e':
        plaintext = input("Enter the plaintext message: ")
        encrypted_message = encrypt_polybius(plaintext, polybius_grid)
        print("Encrypted message:", encrypted_message)
    elif mode == 'd':
        ciphertext = input("Enter the ciphertext message: ")
        decrypted_message = decrypt_polybius(ciphertext, reverse_grid)
        print("Decrypted message:", decrypted_message)
    else:
        print("Invalid option. Please choose 'e' to encrypt or 'd' to decrypt.")

if __name__ == "__main__":
    main()